src/overlays/actors/ovl_Object_Kankyo/z_object_kankyo.c

```diff
@@ -32,41 +57,879 @@ const ActorInit Object_Kankyo_InitVars = {
+            case 473:
+                func_800788CC(NA_SE_VO_NA_HELLO_3);
+                break;
+
+            case 583:
+                func_800F4524(&D_801333D4, NA_SE_VO_NA_HELLO_2, 32);
+                break;
+
+            case 763:
+                func_80078884(NA_SE_EV_NAVY_CRASH - SFX_FLAG);
+                break;
+
+            case 771:
+                func_80078884(NA_SE_VO_RT_THROW);
+                break;
+        }
+    }
+
+    if (globalCtx->envCtx.unk_EE[3] < 64 &&
+        (gSaveContext.entranceIndex != 0x00EE || gSaveContext.sceneSetupIndex != 4 || globalCtx->envCtx.unk_EE[3])) {
+        globalCtx->envCtx.unk_EE[3] += 16;
+    }
+
+    for (i = 0; i < globalCtx->envCtx.unk_EE[3]; i++) {
+        dx = globalCtx->view.lookAt.x - globalCtx->view.eye.x;
+        dy = globalCtx->view.lookAt.y - globalCtx->view.eye.y;
+        dz = globalCtx->view.lookAt.z - globalCtx->view.eye.z;
+        dist = sqrtf(SQ(dx) + SQ(dy) + SQ(dz));
+
+        offset.x = dx / dist;
+        offset.y = dy / dist;
+        offset.z = dz / dist;
+
+        // :O
+        offsetX = offset.x;
+        offsetY = offset.y;
+        offsetZ = offset.z;
+
+        switch (this->effects[i].state) {
+            case 0: // init
+                this->effects[i].base.x = globalCtx->view.eye.x + offsetX * 80.0f;
+                this->effects[i].base.y = globalCtx->view.eye.y + offsetY * 80.0f;
+                this->effects[i].base.z = globalCtx->view.eye.z + offsetZ * 80.0f;
+
+                this->effects[i].pos.x = (Rand_ZeroOne() - 0.5f) * 160.0f;
+                this->effects[i].pos.y = 30.0f;
+                this->effects[i].pos.z = (Rand_ZeroOne() - 0.5f) * 160.0f;
+                this->effects[i].targetSpeed = Rand_ZeroOne() * 1.6f + 0.5f;
+                this->effects[i].alpha = 0;
+                this->effects[i].alphaTimer = Rand_ZeroOne() * 65535.0f;
+                this->effects[i].size = 0.1f;
+                this->effects[i].dirPhase.x = Rand_ZeroOne() * 360.0f;
+                this->effects[i].dirPhase.y = Rand_ZeroOne() * 360.0f;
+                this->effects[i].dirPhase.z = Rand_ZeroOne() * 360.0f;
+                this->effects[i].state++;
+                this->effects[i].timer = 0;
+                break;
+
+            case 1: // blinking fairies / inactive fairy trails
+            case 2: // fairy trails
+                this->effects[i].alphaTimer++;
+                baseX = globalCtx->view.eye.x + offset.x * 80.0f;
+                baseY = globalCtx->view.eye.y + offset.y * 80.0f;
+                baseZ = globalCtx->view.eye.z + offset.z * 80.0f;
+
+                this->effects[i].lastPos.x = this->effects[i].pos.x;
+                this->effects[i].lastPos.y = this->effects[i].pos.y;
+                this->effects[i].lastPos.z = this->effects[i].pos.z;
+
+                playerMoved = true;
+                if (player->actor.velocity.x + player->actor.velocity.y + player->actor.velocity.z == -4.0f) {
```

Comment:
my guess is that y velocity is always -4 on the ground, either because of gravity or just a hardcoded value

Player_UpdateCommon -> func_80847BA0 -> Actor_UpdateBgCheckInfo -> func_8002E2AC
func_8002E2AC seems to do this
```c
            } else if ((arg3 & 0x8) && (actor->gravity < 0.0f)) {
                actor->velocity.y = -4.0f;
            } else {
```